File: repository/src/main/scala/by/rekuts/travelagency/repository/TourSpecification.scala
package by.rekuts.travelagency.repository

import java.time.LocalDate
import java.util

import by.rekuts.travelagency.domain.TourType.TourType
import by.rekuts.travelagency.domain.{Hotel, Tour}
import javax.persistence.criteria.{CriteriaBuilder, Join, Predicate, Root}

class TourSpecification extends Specification[Tour] {
  def this(id: Int) {
    this()
    this.id = id
  }

  var countryId: Int = _
  var date: LocalDate = _
  var duration: Int = _
  var minCost: BigDecimal = _
  var maxCost: BigDecimal = _
  var tourType: TourType = _
  var stars: Int = _
  var userId: Int = _

  override def getPredicates(root: Root[Tour], builder: CriteriaBuilder): util.List[Predicate] = {
    var predicates = super.getPredicates(root, builder)
    if (countryId != 0) {
      predicates.add(
        builder.equal(root.get("country"), countryId)
      )
    }
    if (date != null) {
      predicates.add(
        builder.equal(root.get("date"), date)
      )
    }
    if (duration != 0) {
      predicates.add(
        builder.equal(root.get("duration"), duration)
      )
    }
    if (minCost != null) {
      predicates.add(
        builder.ge(root.get("cost"), minCost)
      )
    }
    if (maxCost != null) {
      predicates.add(
        builder.le(root.get("cost"), maxCost)
      )
    }
    if (tourType != null) {
      predicates.add(
        builder.equal(root.get("tourType"), tourType)
      )
    }
    if (stars != 0) {
      val hotelJoin: Join[Tour, Hotel] = root.join("hotel")
      predicates.add(
        builder.equal(hotelJoin.get("stars"), stars)
      )
    }
    if (userId != 0) {
      predicates.add(
        builder.equal(root.join("users"), userId)
      )
    }
    predicates
  }
}

File: repository/src/main/java/by/rekuts/travelagency/repository/TourSpecification.java
package by.rekuts.travelagency.repository;

import by.rekuts.travelagency.domain.Hotel;
import by.rekuts.travelagency.domain.Tour;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.jpa.domain.Specification;

import javax.persistence.criteria.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Getter @Setter
@NoArgsConstructor
public class TourSpecification implements Specification<Tour> {

    private Integer id;
    private Integer countryId;
    private LocalDate date;
    private Integer duration;
    private BigDecimal minCost;
    private BigDecimal maxCost;
    private Tour.TourType tourType;
    private Integer stars;
    private Integer userId;

    public TourSpecification(Integer id) {
        this.id = id;
    }

    public TourSpecification(Integer id, Integer countryId, LocalDate date, Integer duration,
                             BigDecimal minCost, BigDecimal maxCost, Tour.TourType tourType,
                             Integer stars, Integer userId) {
        this.id = id;
        this.countryId = countryId;
        this.date = date;
        this.duration = duration;
        this.minCost = minCost;
        this.maxCost = maxCost;
        this.tourType = tourType;
        this.stars = stars;
        this.userId = userId;
    }

    @Override
    public Predicate toPredicate(Root<Tour> root, CriteriaQuery<?> query, CriteriaBuilder builder) {
        List<Predicate> predicates = new ArrayList<>();

        if (id != null) {
            predicates.add(builder.equal(root.get("id"), id));
        }
        if (countryId != null) {
            predicates.add(builder.equal(root.get("country"), countryId));
        }
        if (date != null) {
            predicates.add(builder.equal(root.get("date"), date));
        }
        if (duration != null) {
            predicates.add(builder.equal(root.get("duration"), duration));
        }
        if (minCost != null) {
            predicates.add(builder.ge(root.get("cost"), minCost));
        }
        if (maxCost != null) {
            predicates.add(builder.le(root.get("cost"), maxCost));
        }
        if (tourType != null) {
            predicates.add(builder.equal(root.get("tourType"), tourType));
        }
        if (stars != null) {
            Join<Tour, Hotel> hotelJoin = root.join("hotel");
            predicates.add(builder.equal(hotelJoin.get("stars"), stars));
        }
        if (userId != null) {
            predicates.add(builder.equal(root.join("users"), userId));
        }

        // Combine predicates into a single Predicate
        return builder.and(predicates.toArray(new Predicate[0]));
    }
}
